[FLINK-5815] Add resource files configuration for Yarn Mode

wenlong88 opened the pull request:
This PR add three common resource configuration options to yarn mode, which allow user to set single file resource from both local filesystem and remote hdfs filesystem as what we can do using mapreduce, including:
1. add -yfiles . -ylibjars for adding local resource file to yarn per-job cluster, user can provide a list of file paths to add some local jars or dictionary files to yarn distributed cache.
2. add -yarchives for adding remote resource files to yarn per-job cluster, user can provide a list of uri of files which can be stored on hdfs, user can rename the file by fragment of the uri.
all of the files will be distributed to every TM and JM by yarn and added to classpath of TM and JM.

tillrohrmann commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/FileUtilsTest.java:
```diff
@@ -0,0 +1,117 @@
+	@Test
+	public void testCompareFs() throws IOException {
+		File localDir = temporaryFolder.newFolder();
+		Assert.assertFalse(FileUtils.compareFs(new Path(localDir.toURI()).getFileSystem(), FS));
```
For this test, I think we don't have to create a HDFSCluster to obtain an HDFS `FileSystem`. Can't we create the respective URIs?

wenlong88 replied:
Just taking advantage of the hdfs we have created for localize file.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/util/FileUtils.java:
```diff
@@ -258,7 +261,82 @@ public static boolean deletePathIfEmpty(FileSystem fileSystem, Path path) throws
+			fragment = originalPath.getName();
+		}
+		Path newPath = new Path(localDir, fragment);
+		IOUtils.copyBytes(remoteFs.open(originalPath), localFs.create(newPath, FileSystem.WriteMode.OVERWRITE), true);
```
Maybe we could issue a warning if we try to overwrite a file?

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -248,6 +253,14 @@ public void addShipFiles(List<File> shipFiles) {
+		try {
+			this.archives.add(new URI(archive));
+		} catch (URISyntaxException e) {
+			throw new IllegalArgumentException(e);
```
Maybe we could add a short exception message.

wenlong88 replied:
fixed

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -742,6 +755,24 @@ public FileVisitResult preVisitDirectory(java.nio.file.Path dir, BasicFileAttrib
+				FileUtils.localizeRemoteFiles(new org.apache.flink.core.fs.Path(remoteParent.toUri()), originURI).toUri());
+			URI targetURI = target.toUri();
+			if (targetURI.getFragment() == null && fragment != null) {
+				targetURI = new URI(target.toUri().toString() + "#" + fragment);
```
no need for `toString`

tillrohrmann replied:
Why do we need to append a fragment here? I thought if `remoteParent` contained a fragment, then `target` would be a file of that fragment and be named after the fragment.

tillrohrmann replied:
Haven't we already extracted the target URI in `targetURI`?

wenlong88 replied:
you are right about the file naming rule. When using Path as localizeFile's output type, the uri fragment will be dropped, so we need to recover the fragment there. I have change the return type to `URI`, so it is not needed any more.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/util/FileUtils.java:
```diff
@@ -258,7 +261,82 @@ public static boolean deletePathIfEmpty(FileSystem fileSystem, Path path) throws
+	 * @return
+	 * @throws IOException
+	 */
+	public static Path localizeRemoteFiles(Path localDir, URI originalFile) throws IOException {
```
Isn't it better called `localizeRemoteFile`, because it only copies one file?

wenlong88 replied:
fixed

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -742,6 +755,24 @@ public FileVisitResult preVisitDirectory(java.nio.file.Path dir, BasicFileAttrib
+			Path remoteParent = Utils.getRemoteResourceRoot(appId.toString(), new Path(originURI.getPath()), fs.getHomeDirectory());
+			String fragment = originURI.getFragment();
+			Path target = new Path(
+				FileUtils.localizeRemoteFiles(new org.apache.flink.core.fs.Path(remoteParent.toUri()), originURI).toUri());
```
Maybe `FileUtils.localizeRemoteFiles` should return a `URI` if the returned `Path` has to be unpacked and then again packed into another `Path` type.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -742,6 +755,24 @@ public FileVisitResult preVisitDirectory(java.nio.file.Path dir, BasicFileAttrib
+			LocalResource archive = Records.newRecord(LocalResource.class);
+			FileStatus state = fs.getFileStatus(target);
+			Utils.registerLocalResource(targetURI, state.getLen(), state.getModificationTime(), archive);
+			localResources.put(fragment, archive);
```
What if `fragment == null`? Or is `fragment` always not `null`?

wenlong88 replied:
fixed.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -742,6 +755,24 @@ public FileVisitResult preVisitDirectory(java.nio.file.Path dir, BasicFileAttrib
+			FileStatus state = fs.getFileStatus(target);
+			Utils.registerLocalResource(targetURI, state.getLen(), state.getModificationTime(), archive);
+			localResources.put(fragment, archive);
+			paths.add(new Path(targetURI));
```
isn't that the same as `target`?

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -742,6 +755,24 @@ public FileVisitResult preVisitDirectory(java.nio.file.Path dir, BasicFileAttrib
+			paths.add(new Path(targetURI));
+			classPathBuilder.append(fragment).append(File.pathSeparator);
+			envShipFileList.append(targetURI).append(",");
+		}
```
I've got the feeling that this code snippet contains a lot of path to URI and back to path conversions. Maybe we could streamline it a little bit to avoid unnecessary object instantiations.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -136,21 +137,36 @@ public static Path setupLocalResource(
 	}
 
+	public static Path getRemoteResourceRoot(
+			String appId, Path localRsrcPath,
```
every parameter in a separate line

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -136,21 +137,36 @@ public static Path setupLocalResource(
 
+	public static Path getRemoteResourceRoot(
+			String appId, Path localRsrcPath,
+			Path homedir) throws IOException {
```
isn't `homedir` better called `targetDir`?

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -136,21 +137,36 @@ public static Path setupLocalResource(
 		return dst;
 	}
 
+	public static Path getRemoteResourceRoot(
```
Maybe it's better called `getRemoteResourcePath`

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -131,6 +134,8 @@
 	/** Lazily initialized list of files to ship */
 	protected List<File> shipFiles = new LinkedList<>();
 
+	private List<URI> archives = new LinkedList<>();
```
`LinkedLists` are not so fast. Better to use an `ArrayList` and initialize it with an initial capacity.

wenlong88 replied:
it is because we cannot know the size in advance, and there are usually little elements for archives, so I just keep the same structure with shipFiles.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnCLI.java:
```diff
@@ -85,6 +89,10 @@ public FlinkYarnCLI(String shortPrefix, String longPrefix) {
 
+		LIB_JARS = new Option(shortPrefix + "libjars", longPrefix + "libjars", true, "Jar file paths for job, like /home/user/lib/test.jar");
+		FILES = new Option(shortPrefix + "files", longPrefix + "files", true, "Normal file paths for job, like /home/user/lib/test.dict");
+		ARCHIVES = new Option(shortPrefix + "archives", longPrefix + "archives", true, "Archived file uris for job, like hdfs:///users/flink/common_dict#dict");
```
I'm wondering whether we shouldn't offer these options also for the standalone and mesos case. We could use the BlobStore to distribute the files.

wenlong88 replied:
It is totally different using the BlobStore. For clusters with BlobStore, DistributedCache may be enough. I think in the future, we can support DistributedCache with BlobStore in the future, and support external BlobStore structure which already run on existed framework, like yarn distributed cache in the future

wenlong88 commented on flink-core/src/main/java/org/apache/flink/util/FileUtils.java:
```diff
@@ -258,7 +261,82 @@ public static boolean deletePathIfEmpty(FileSystem fileSystem, Path path) throws
+	 * @param destFs
+	 * @return
+	 */
+	public static boolean compareFs(FileSystem srcFs, FileSystem destFs) {
```
Here I just want to check whether the file system is the same or not, if we provide a function compare two URIs, we need to compare more extra elements.

wenlong88 commented on flink-core/src/main/java/org/apache/flink/util/FileUtils.java:
```diff
@@ -258,7 +261,82 @@ public static boolean deletePathIfEmpty(FileSystem fileSystem, Path path) throws
+		URI dstUri = destFs.getUri();
+
+		// check schema
+		if (srcUri.getScheme() == null) {
```
fixed

wenlong88 commented on flink-core/src/main/java/org/apache/flink/util/FileUtils.java:
```diff
@@ -258,7 +261,82 @@ public static boolean deletePathIfEmpty(FileSystem fileSystem, Path path) throws
+			if (!srcHost.equals(dstHost)) {
+				try {
+					srcHost = InetAddress.getByName(srcHost).getCanonicalHostName();
+					dstHost = InetAddress.getByName(dstHost).getCanonicalHostName();
```
I think FQDN is better when multiple service running on the same machine, what do yo think?

wenlong88 commented on flink-core/src/main/java/org/apache/flink/util/FileUtils.java:
```diff
@@ -258,7 +261,82 @@ public static boolean deletePathIfEmpty(FileSystem fileSystem, Path path) throws
+			return false;
+		} else if (srcHost != null && dstHost == null) {
+			return false;
+		}
```
ok, fixing this.

wenlong88 commented on flink-core/src/main/java/org/apache/flink/util/FileUtils.java:
```diff
@@ -258,7 +261,82 @@ public static boolean deletePathIfEmpty(FileSystem fileSystem, Path path) throws
+	 * @param localDir      local directory to store remote files.
+	 * @param originalFile  original file to check
+	 * @return
+	 * @throws IOException
```
fixed

wenlong88 commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/FileUtilsTest.java:
```diff
@@ -0,0 +1,117 @@
+import static org.junit.Assert.fail;
+
+/**
+ * Created by wenlong.lwl on 2017/2/21.
```
sorry, this is a missing clean up from our code base.

wenlong88 commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/FileUtilsTest.java:
```diff
@@ -0,0 +1,117 @@
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+			fail("Could not create HDFS mini cluster " + e.getMessage());
```
ok

wenlong88 commented:
@tillrohrmann sorry for late response, I have addressed the comments. Thinks for the review.

Currently for standalone, and mesos, there are already some ways to do things like that, `--classpath `, `DistributedCache in batch` , we need to do some refactor and make all of this more clear to user in the future before adding a new one. But it is necessary to make flip-6 implementation in yarn run well.

the test is not added because the `flink-yarn-test` is disabled, and I was not able to enable it. Is there any problem in running yarn test now?

EronWright commented:
Just a note: For Mesos, inevitably we'll add support for arbitrary URLs to be downloaded by the container.   The Mesos fetcher deals with HTTP URLs primarily but actually supports HDFS URLs too (if configured; seems rarely used).

Stepping back, we need the clustering layer to converge because supporting these features in a one-off way is becoming unsustainable.

tillrohrmann commented:
Closing this PR because it is outdated.
